File: tools/librapp/textmatch.py
"""Deciding whether two records name the same book or the same person.

Every source spells things differently. A store export writes 'Terry
Pratchett', a spreadsheet writes 'Pratchett, Terry', a photograph shows only
what fits on a spine. Nothing here is about a particular source; it is the
shared vocabulary the merge uses to reconcile them.

Folded forms are for comparison only. Display always uses the original.
"""

from __future__ import annotations

import re
import unicodedata
from difflib import SequenceMatcher

# Title similarity above this counts as the same book, given the authors
# already match. Tuned so that edition noise ('(Biblioteca Clasica Gredos n 387)')
# does not separate a book from itself.
TITLE_MATCH_THRESHOLD = 0.62

# Where no author is known on either side, the title carries the match alone
# and has to be all but identical.
UNCREDITED_MATCH_THRESHOLD = 0.95

# Parenthetical or trailing edition noise, stripped before comparison only.
EDITION_NOISE = re.compile(
    r"\((?:[^()]*\b(?:biblioteca|cl[aá]sica|gredos|bolsillo|edici[oó]n|ed\.|"
    r"vol\.|n[oº]?\s*\d+|divulgaci[oó]n|runas|bloomsbury|sigma)\b[^()]*)\)",
    re.IGNORECASE,
)

# Series stated inside the title, e.g. '(Los casos del Departamento Q 1)' or
# 'La Primera Ley: Libro II'. Deliberately conservative: anything not matching
# these shapes is left for review rather than invented.
SERIES_PATTERNS = [
    re.compile(r"\((?P<series>[^()]+?)\s+(?P<index>\d{1,2})\)\s*$"),
    re.compile(r"\((?P<series>[^()]+?)\s+(?P<index>[IVX]{1,5})\)\s*$"),
    re.compile(r":\s*(?P<series>[^:]+?):\s*Libro\s+(?P<index>[IVX]{1,5}|\d{1,2})\b"),
]
ROMAN = {"I": 1, "II": 2, "III": 3, "IV": 4, "V": 5, "VI": 6, "VII": 7,
         "VIII": 8, "IX": 9, "X": 10, "XI": 11, "XII": 12}


def strip_accents(s: str) -> str:
    return "".join(c for c in unicodedata.normalize("NFD", s) if not unicodedata.combining(c))


def clean(s: str) -> str:
    """Whitespace and unicode tidied, but the text otherwise untouched.

    Safe for display: NFKC also turns the non-breaking spaces some sources hide
    inside words into ordinary ones.
    """
    return re.sub(r"\s+", " ", unicodedata.normalize("NFKC", s)).strip()


def fold(s: str) -> str:
    """Aggressive normalisation for comparison keys only, never for display."""
    s = strip_accents(unicodedata.normalize("NFKC", s)).casefold()
    s = re.sub(r"[^\w\s]", " ", s)
    return re.sub(r"\s+", " ", s).strip()


def author_tokens(name: str) -> frozenset[str]:
    """Comparable token set for a personal name.

    Sources disagree on order and punctuation: 'Pratchett, Terry' against
    'Terry Pratchett'. A set ignores order. Bare initials ('H. P.') carry no
    signal and are dropped, leaving the surname to do the work.
    """
    return frozenset(t for t in fold(name.replace(",", " ")).split() if len(t) > 1)


def split_credits(name: str) -> list[str]:
    """Split a multi-author credit into individual names.

    Semicolons separate co-authors ('Herbert, Frank; Ransom, Bill'); commas
    cannot, because they already mean the surname-first inversion.
    """
    parts = [p.strip() for p in re.split(r";|\s+&\s+|\s+y\s+otros\b", name) if p.strip()]
    return parts or ([name.strip()] if name.strip() else [])


# A bracketed stand-in where a person's name would go: '[Anonimo]', '[Varios]',
# '[Reference]'. It labels the entry rather than naming anybody.
LABEL = re.compile(r"\[([^\]]+)\]")

# A parenthetical crediting whoever prepared the edition rather than wrote it.
EDITORIAL = re.compile(
    r"\((?:\s*(?:eds?\.|edited by|trans\.|translated by|coord\.|comp\.)\s*)(?P<who>[^)]*)\)",
    re.IGNORECASE,
)


def credits_and_label(field: str) -> tuple[list[str], str | None]:
    """Pull the people out of a hand-written author field.

    Hand-kept lists put more than a name in that column: a stand-in for an
    anonymous work, the editors of an edition, a real name behind a pen name.
    Left alone these become authors called '[Varios] (eds. Meyer' - which then
    merge with real people and corrupt the author list - so they are separated
    here, where the mess is, rather than defended against everywhere else.

    Returns the personal names, and any bracketed label that stood in for one.
    """
    field = clean(field)
    if not field:
        return [], None

    label = None
    match = LABEL.search(field)
    if match:
        label = match.group(1).strip()
        field = LABEL.sub(" ", field)

    people: list[str] = []
    for editorial in EDITORIAL.finditer(field):
        # Inside an editorial credit a comma separates people, not surname
        # from forename, so it is safe to split on here and nowhere else.
        people += [p.strip() for p in re.split(r",|\s+&\s+|;", editorial.group("who")) if p.strip()]
    field = EDITORIAL.sub(" ", field)

    # Anything left in brackets or parentheses is an aside, not a name.
    field = re.sub(r"\([^)]*\)", " ", field)
    people = [p for p in (clean(x) for x in split_credits(field)) if p] + people
    return [p for p in people if author_tokens(p)], label


def index_keys(name: str) -> set[str]:
    """Keys a name is filed under when looking for candidate matches.

    Includes a five-character prefix of every token, so spellings differing
    only in their tail - 'Platon' against 'Plato' - still meet. Loose indexing
    is safe: it only proposes candidates, and the title comparison decides.
    """
    keys: set[str] = set()
    for credit in split_credits(name):
        for token in author_tokens(credit):
            keys.add(token)
            keys.add(token[:5])
    return keys


def title_key(title: str) -> str:
    """Title reduced for comparison: edition noise and parentheses removed."""
    t = EDITION_NOISE.sub(" ", title)
    return fold(re.sub(r"\([^()]*\)", " ", t))


def title_head(title: str) -> str:
    """Just the main title, with any subtitle after a colon or dash dropped.

    Sources disagree about subtitles far more than about titles: a spreadsheet
    records 'Los tonicos de la voluntad' where a store carries the full
    'Los tonicos de la voluntad: Reglas y consejos sobre investigacion...'.
    """
    return title_key(re.split(r"[:–—]|\s-\s", title, maxsplit=1)[0])


def similar(a: str, b: str) -> float:
    return SequenceMatcher(None, a, b).ratio()


def title_score(a: str, b: str) -> float:
    """How strongly two titles denote the same book.

    Containment matters more than edit distance here. One source routinely
    holds a longer form of the other - a subtitle, an edition line, a series
    note - which wrecks a plain ratio while being near-proof of identity.
    """
    if not a or not b:
        return 0.0
    if a == b:
        return 1.0
    short, long = sorted((a, b), key=len)
    # Whole-word containment only: 'obras' inside 'obras de howard phillips
    # lovecraft' is the same book, but inside 'obrass' it would be noise.
    if len(short) >= 5 and f" {short} " in f" {long} ":
        return 0.97 if long.startswith(short) else 0.90
    return similar(a, b)


def best_title_score(a: str, b: str) -> float:
    """Title agreement, judged on the full titles and on the main titles."""
    return max(title_score(title_key(a), title_key(b)),
               title_score(title_head(a), title_head(b)))


def slugify(*parts: str) -> str:
    joined = "-".join(p for p in parts if p)
    return re.sub(r"[^\w]+", "-", strip_accents(joined).casefold()).strip("-")[:80] or "untitled"


def detect_series(title: str) -> tuple[str | None, int | None]:
    """Series name and volume number, when the title states them plainly."""
    for pattern in SERIES_PATTERNS:
        m = pattern.search(title)
        if not m:
            continue
        raw_index = m.group("index").upper()
        index = int(raw_index) if raw_index.isdigit() else ROMAN.get(raw_index)
        if index is None:
            continue
        series = re.sub(r"\s+", " ", m.group("series")).strip(" :,-")
        # Publisher collections number themselves 'Solaris ficcion no 12',
        # leaving a dangling 'no' once the number is taken as the index.
        series = re.sub(r"\s+n[oº°]?\.?$", "", series, flags=re.IGNORECASE).strip(" :,-")
        if len(series) < 3:
            continue
        return series, index
    return None, None

File: web/src/core/textmatch.js
// Port of tools/librapp/textmatch.py. Deciding whether two records name the
// same book or the same person.
//
// Two differences between Python and JavaScript regular expressions matter
// here, and both are silent if you get them wrong:
//
//   * Python's `\w` and `\b` are Unicode-aware. JavaScript's are ASCII-only, so
//     /\bclasica\b/ behaves one way and /\bclásica\b/ another. Every word
//     boundary below is written as an explicit lookaround over WORD instead.
//   * Python's `re.split(..., maxsplit=1)` stops splitting; JavaScript's
//     `split(re, limit)` splits everywhere and then truncates. `splitFirst`
//     does what Python does.

import { ratio } from './difflib.js'

/** What Python's `\w` matches under Unicode: letters, digits, underscore. */
const WORD = '\\p{L}\\p{N}_'
const notWord = (body) => new RegExp(`(?<![${WORD}])(?:${body})(?![${WORD}])`, 'iu')

export const TITLE_MATCH_THRESHOLD = 0.62
export const UNCREDITED_MATCH_THRESHOLD = 0.95

const EDITION_WORDS =
  'biblioteca|cl[aá]sica|gredos|bolsillo|edici[oó]n|ed\\.|vol\\.|n[oº]?\\s*\\d+|' +
  'divulgaci[oó]n|runas|bloomsbury|sigma'
const EDITION_NOISE = new RegExp(
  `\\((?:[^()]*(?<![${WORD}])(?:${EDITION_WORDS})(?![${WORD}])[^()]*)\\)`,
  'giu',
)

const SERIES_PATTERNS = [
  /\((?<series>[^()]+?)\s+(?<index>\d{1,2})\)\s*$/u,
  /\((?<series>[^()]+?)\s+(?<index>[IVX]{1,5})\)\s*$/u,
  new RegExp(`:\\s*(?<series>[^:]+?):\\s*Libro\\s+(?<index>[IVX]{1,5}|\\d{1,2})(?![${WORD}])`, 'u'),
]
const ROMAN = { I: 1, II: 2, III: 3, IV: 4, V: 5, VI: 6, VII: 7, VIII: 8, IX: 9, X: 10, XI: 11, XII: 12 }

const LABEL = /\[([^\]]+)\]/u
const EDITORIAL =
  /\((?:\s*(?:eds?\.|edited by|trans\.|translated by|coord\.|comp\.)\s*)(?<who>[^)]*)\)/giu

// Python folds case with str.casefold(), which is not str.lower(). Where they
// differ, one letter folds to several or to a different letter entirely, and
// JavaScript's toLowerCase() leaves it alone. That silently breaks matching:
// 'Straße' and 'STRASSE' stop being the same word, and a Greek title ending in
// a final sigma stops matching the same title written with a medial one.
//
// Only the cases that can plausibly appear in a book catalogue are listed —
// German, Greek, and the Latin ligatures. The ligatures matter even though
// NFKC resolves them, because slugify() folds without normalising first.
const CASEFOLD_SPECIALS = new Map([
  ['ß', 'ss'], // ß
  ['ẞ', 'ss'], // ẞ
  ['ς', 'σ'], // final sigma folds to sigma
  ['ŉ', 'ʼn'], // ŉ
  ['ǰ', 'ǰ'], // ǰ
  ['İ', 'i̇'], // İ
  ['ẖ', 'ẖ'],
  ['ẗ', 'ẗ'],
  ['ẘ', 'ẘ'],
  ['ẙ', 'ẙ'],
  ['ẚ', 'aʾ'],
  ['ﬀ', 'ff'],
  ['ﬁ', 'fi'],
  ['ﬂ', 'fl'],
  ['ﬃ', 'ffi'],
  ['ﬄ', 'ffl'],
  ['ﬅ', 'st'],
  ['ﬆ', 'st'],
])

/** Python's str.casefold(), for the letters a book catalogue can contain. */
export function casefold(s) {
  let out = ''
  for (const ch of (s || '').toLowerCase()) out += CASEFOLD_SPECIALS.get(ch) ?? ch
  return out
}

export const stripAccents = (s) => (s || '').normalize('NFD').replace(/\p{M}/gu, '')

/** Whitespace and Unicode tidied, text otherwise untouched. Safe for display. */
export const clean = (s) => (s || '').normalize('NFKC').replace(/\s+/g, ' ').trim()

/** Aggressive normalisation for comparison keys only, never for display. */
export const fold = (s) =>
  casefold(stripAccents((s || '').normalize('NFKC')))
    .replace(new RegExp(`[^${WORD}\\s]`, 'gu'), ' ')
    .replace(/\s+/g, ' ')
    .trim()

/** Python's re.split(pattern, s, maxsplit=1)[0]. */
function splitFirst(s, pattern) {
  const re = new RegExp(pattern.source, pattern.flags.includes('g') ? pattern.flags : pattern.flags + 'g')
  const m = re.exec(s)
  return m ? s.slice(0, m.index) : s
}

/**
 * Comparable token set for a personal name.
 *
 * Sources disagree on order and punctuation: 'Pratchett, Terry' against
 * 'Terry Pratchett'. A set ignores order. Bare initials ('H. P.') carry no
 * signal and are dropped, leaving the surname to do the work.
 */
export function authorTokens(name) {
  return new Set(
    fold((name || '').replace(/,/g, ' '))
      .split(' ')
      .filter((t) => t.length > 1),
  )
}

/** A stable key for a token set, so it can be used in a Map. */
export const tokenKey = (tokens) => [...tokens].sort().join(' ')

/**
 * Split a multi-author credit into individual names. Semicolons separate
 * co-authors; commas cannot, because they already mean surname-first.
 */
export function splitCredits(name) {
  const parts = (name || '')
    .split(new RegExp(`;|\\s+&\\s+|\\s+y\\s+otros(?![${WORD}])`, 'u'))
    .map((p) => p.trim())
    .filter(Boolean)
  if (parts.length) return parts
  const whole = (name || '').trim()
  return whole ? [whole] : []
}

/**
 * Keys a name is filed under when looking for candidate matches. Includes a
 * five-character prefix of every token, so spellings differing only in their
 * tail — 'Platon' against 'Plato' — still meet. Loose indexing is safe: it only
 * proposes candidates, and the title comparison decides.
 */
export function indexKeys(name) {
  const keys = new Set()
  for (const credit of splitCredits(name)) {
    for (const token of authorTokens(credit)) {
      keys.add(token)
      keys.add(token.slice(0, 5))
    }
  }
  return keys
}

/** Title reduced for comparison: edition noise and parentheses removed. */
export const titleKey = (title) =>
  fold((title || '').replace(EDITION_NOISE, ' ').replace(/\([^()]*\)/gu, ' '))

/** Just the main title, with any subtitle after a colon or dash dropped. */
export const titleHead = (title) => titleKey(splitFirst(title || '', /[:–—]|\s-\s/u))

export const similar = ratio

/**
 * How strongly two titles denote the same book.
 *
 * Containment matters more than edit distance: one source routinely holds a
 * longer form of the other — a subtitle, an edition line, a series note — which
 * wrecks a plain ratio while being near-proof of identity.
 */
export function titleScore(a, b) {
  if (!a || !b) return 0.0
  if (a === b) return 1.0
  const [short, long] = a.length <= b.length ? [a, b] : [b, a]
  if (short.length >= 5 && ` ${long} `.includes(` ${short} `)) {
    return long.startsWith(short) ? 0.97 : 0.9
  }
  return ratio(a, b)
}

/** Title agreement, judged on the full titles and on the main titles. */
export const bestTitleScore = (a, b) =>
  Math.max(titleScore(titleKey(a), titleKey(b)), titleScore(titleHead(a), titleHead(b)))

export function slugify(...parts) {
  const joined = parts.filter(Boolean).join('-')
  const slug = casefold(stripAccents(joined))
    .replace(new RegExp(`[^${WORD}]+`, 'gu'), '-')
    .replace(/^-+|-+$/g, '')
    .slice(0, 80)
  return slug || 'untitled'
}

/** Series name and volume number, when the title states them plainly. */
export function detectSeries(title) {
  for (const pattern of SERIES_PATTERNS) {
    const m = pattern.exec(title || '')
    if (!m) continue
    const raw = m.groups.index.toUpperCase()
    const index = /^\d+$/.test(raw) ? Number(raw) : ROMAN[raw]
    if (index === undefined) continue
    let series = m.groups.series.replace(/\s+/g, ' ').replace(/^[\s:,-]+|[\s:,-]+$/g, '')
    // Publisher collections number themselves 'Solaris ficcion no 12', leaving
    // a dangling 'no' once the number is taken as the index.
    series = series
      .replace(/\s+n[oº°]?\.?$/iu, '')
      .replace(/^[\s:,-]+|[\s:,-]+$/g, '')
    if (series.length < 3) continue
    return [series, index]
  }
  return [null, null]
}

/**
 * Pull the people out of a hand-written author field.
 *
 * Hand-kept lists put more than a name in that column: a stand-in for an
 * anonymous work, the editors of an edition, a real name behind a pen name.
 * Left alone these become authors called '[Varios] (eds. Meyer' — which then
 * merge with real people and corrupt the author list.
 *
 * Returns [people, label].
 */
export function creditsAndLabel(field) {
  let text = clean(field)
  if (!text) return [[], null]

  let label = null
  const m = LABEL.exec(text)
  if (m) {
    label = m[1].trim()
    text = text.replace(new RegExp(LABEL.source, 'gu'), ' ')
  }

  const people = []
  for (const editorial of text.matchAll(EDITORIAL)) {
    // Inside an editorial credit a comma separates people, not surname from
    // forename, so it is safe to split on here and nowhere else.
    for (const p of editorial.groups.who.split(/,|\s+&\s+|;/u)) {
      if (p.trim()) people.push(p.trim())
    }
  }
  text = text.replace(EDITORIAL, ' ').replace(/\([^)]*\)/gu, ' ')

  const found = splitCredits(text)
    .map((x) => clean(x))
    .filter(Boolean)
    .concat(people)
  return [found.filter((p) => authorTokens(p).size > 0), label]
}
